explore/exploration.h

```diff
@@ -0,0 +1,180 @@
+  {
+    typedef typename std::iterator_traits<InputIt>::iterator_category scores_category;
+    typedef typename std::iterator_traits<OutputIt>::iterator_category pdf_category;
+
+    generate_softmax(lambda, scores_begin, scores_last, scores_category(), pdf_first, pdf_last, pdf_category());
+  }
+
+  template<typename InputIt, typename OutputIt>
+  void generate_bag(InputIt top_actions_begin, InputIt top_actions_last, std::input_iterator_tag top_actions_tag, OutputIt pdf_first, OutputIt pdf_last, std::random_access_iterator_tag pdf_tag)
+  {
+    if (pdf_first == pdf_last)
+      return;
+
+    uint32_t num_models = std::accumulate(top_actions_begin, top_actions_last, 0);
+	if (num_models == 0)
+	{
+	#ifdef EXPLORE_NOEXCEPT
+		*pdf_first = 1;
+		return;
+  #else
+		throw std::out_of_range("must supply at least one top_action from a model");
+  #endif
+	}
+
+    // divide late to improve numeric stability
+    InputIt t_a = top_actions_begin;
+    for (OutputIt d = pdf_first; d != pdf_last && t_a != top_actions_last; ++d, ++t_a)
+      *d = *t_a / (float)num_models;
+  }
+
+  template<typename InputIt, typename OutputIt>
+  void generate_bag(InputIt top_actions_begin, InputIt top_actions_last, OutputIt pdf_first, OutputIt pdf_last)
+  {
+    typedef typename std::iterator_traits<InputIt>::iterator_category top_actions_category;
+    typedef typename std::iterator_traits<OutputIt>::iterator_category pdf_category;
+
+    generate_bag(top_actions_begin, top_actions_last, top_actions_category(), pdf_first, pdf_last, pdf_category());
+  }
+
+  // Note: must be inline to compile on Windows VS2017
+  template<typename It>
+  void enforce_minimum_probability(float min_prob, bool update_zero_elements, It pdf_first, It pdf_last, std::random_access_iterator_tag pdf_tag)
+  {
+	size_t num_actions = pdf_last - pdf_first;
+
+    //input: a probability distribution
+    //output: a probability distribution with all events having probability > min_prob.  This includes events with probability 0 if zeros = true
+    if (min_prob > 0.999) // uniform exploration
+    {
+      size_t support_size = num_actions;
+      if (!update_zero_elements)
+      {
+        for (It d = pdf_first; d != pdf_last; ++d)
+          if (*d == 0)
+            support_size--;
+      }
+
+	  for (It d = pdf_first; d != pdf_last; ++d)
+        if (update_zero_elements || *d > 0)
+          *d = 1.f / support_size;
+
+      return;
+    }
+
+    min_prob /= num_actions;
+    float touched_mass = 0.;
+    float untouched_mass = 0.;
+
+	for (It d = pdf_first; d != pdf_last; ++d)
+    {
+      auto& prob = *d;
+      if ((prob > 0 || (prob == 0 && update_zero_elements)) && prob <= min_prob)
+      {
+        touched_mass += min_prob;
+        prob = min_prob;
+      }
+      else
+        untouched_mass += prob;
+    }
+
+    if (touched_mass > 0.)
+    {
+	#ifndef EXPLORE_NOEXCEPT
```

Comment:
Maybe define an alternate semantics instead?

Reply:
If I understand the code correctly, this means that the PDF is not properly normalized. Should everything be re-normalized?
It feels a bit weird that we only re-normalize if just the elements that would be adjusted sum to more than 1.

Reply:
touched_mass large implies that basically everything is not meeting the minimum probability so the function is being asked to make the lower bound to large.  A reasonable alternate semantics in this case is to simply create a uniform distribution over each entry generating touched_mass.